src/metabase/notification/send.clj

```diff
@@ -117,41 +123,59 @@
-                    (log/infof "[Notification %d] Sending message to channel %s"
-                               (:id notification-info) (:channel_type handler))
-                    (channel-send-retrying! (:id notification-info) (:payload_type notification-info) handler message))))
-              (do-after-notification-sent notification-info notification-payload)
-              (log/infof "[Notification %d] Sent successfully" (:id notification-info)))
-            (log/infof "[Notification %d] Skipping" (:id notification-info))))))
-    (catch Exception e
-      (log/errorf e "[Notification %d] Failed to send" (:id notification-info))
-      (throw e)))
-  nil)
+  (u/with-timer-ms [duration-ms-fn]
+    (try
+      (log/infof "[Notification %d] Sending" (:id notification-info))
+      (prometheus/inc! :metabase-notification/concurrent-tasks 1)
+      (let [handlers (noti-handlers notification-info)]
+        (task-history/with-task-history {:task          "notification-send"
+                                         :task_details {:notification_id       (:id notification-info)
+                                                        :notification_handlers (map #(select-keys % [:id :channel_type :channel_id :template_id]) handlers)}}
+          (let [notification-payload (notification.payload/notification-payload notification-info)]
+            (if (notification.payload/should-send-notification? notification-payload)
+              (do
+                (log/debugf "[Notification %d] Found %d handlers" (:id notification-info) (count handlers))
+                (doseq [handler handlers]
+                  (let [channel-type (:channel_type handler)
+                        messages     (channel/render-notification
+                                      channel-type
+                                      notification-payload
+                                      (:template handler)
+                                      (:recipients handler))]
+                    (log/debugf "[Notification %d] Got %d messages for channel %s with template %d"
+                                (:id notification-info) (count messages)
+                                (handler->channel-name handler)
+                                (-> handler :template :id))
+                    (doseq [message messages]
+                      (log/infof "[Notification %d] Sending message to channel %s"
+                                 (:id notification-info) (:channel_type handler))
+                      (channel-send-retrying! (:id notification-info) (:payload_type notification-info) handler message))))
+                (do-after-notification-sent notification-info notification-payload)
+                (log/infof "[Notification %d] Sent successfully" (:id notification-info)))
+              (log/infof "[Notification %d] Skipping" (:id notification-info)))
+            (prometheus/inc! :metabase-notification/send-ok {:payload-type (:payload_type notification-info)} 1))))
```

Comment:
nit: We use the `payload_type` in so many places, I think it's worth hoisting into the surround `let` block.

Reply:
yep, hoised the id too